pkg/jaegerproxy/proxy.go

```diff
@@ -0,0 +1,173 @@
+// This file and its contents are licensed under the Apache License 2.0.
+// Please see the included NOTICE for copyright information and
+// LICENSE for a copy of the license.
+
+package jaegerproxy
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"time"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/jaegertracing/jaeger/proto-gen/storage_v1"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+)
+
+const DefaultTimeout = time.Duration(5 * time.Second)
+
+type ProxyConfig struct {
+	TLS                bool          `yaml:"promscale-tls,omitempty"`
+	CaFile             string        `yaml:"promscale-cafile,omitempty"`
+	ServerAddr         string        `yaml:"promscale-server,omitempty"`
+	ServerHostOverride string        `yaml:"promscale-server-host-override,omitempty"`
+	ConnectTimeout     time.Duration `yaml:"timeout,omitempty"`
+}
+
+type Proxy struct {
+	config                  ProxyConfig
+	logger                  hclog.Logger
+	conn                    *grpc.ClientConn
+	spanReaderClient        storage_v1.SpanReaderPluginClient
+	spanWriterClient        storage_v1.SpanWriterPluginClient
+	dependencyReaderClient  storage_v1.DependenciesReaderPluginClient
+	capClient               storage_v1.PluginCapabilitiesClient
+	archiveSpanReaderClient storage_v1.ArchiveSpanReaderPluginClient
+	archiveSpanWriterClient storage_v1.ArchiveSpanWriterPluginClient
+}
+
+func New(config ProxyConfig, logger hclog.Logger) (*Proxy, error) {
+	var opts []grpc.DialOption
+	var err error
+	if config.TLS {
+		if config.CaFile == "" {
+			return nil, fmt.Errorf("ca file is required with TLS")
+		}
+		creds, err := credentials.NewClientTLSFromFile(config.CaFile, config.ServerHostOverride)
+		if err != nil {
+			log.Fatalf("Failed to create TLS credentials %v", err)
+		}
+		opts = append(opts, grpc.WithTransportCredentials(creds))
+	} else {
+		opts = append(opts, grpc.WithInsecure())
+	}
+
+	if config.ConnectTimeout == 0 {
+		config.ConnectTimeout = DefaultTimeout
+	}
+
+	opts = append(opts, grpc.WithBlock())
+	ctx, cancel := context.WithTimeout(context.Background(), config.ConnectTimeout)
+	defer cancel()
+	conn, err := grpc.DialContext(ctx, config.ServerAddr, opts...)
+	if err != nil {
+		log.Fatalf("fail to dial: %v", err)
+	}
+
+	return &Proxy{
+		config:                  config,
+		logger:                  logger,
+		conn:                    conn,
+		spanReaderClient:        storage_v1.NewSpanReaderPluginClient(conn),
+		dependencyReaderClient:  storage_v1.NewDependenciesReaderPluginClient(conn),
+		capClient:               storage_v1.NewPluginCapabilitiesClient(conn),
+		archiveSpanReaderClient: storage_v1.NewArchiveSpanReaderPluginClient(conn),
+		archiveSpanWriterClient: storage_v1.NewArchiveSpanWriterPluginClient(conn),
+	}, nil
+}
+
+func (p *Proxy) Close() {
+	p.conn.Close()
```

Comment:
I guess we should close all the other clients, like `spanReaderClient`, etc...?

Reply:
Those don't have a Close function